Automattic/wp-calypso
# Initial support for premium domain suggestions

#### Changes proposed in this Pull Request

This change serves as the initial work to support purchase of premium domains in Calypso, though this diff introduces a new `domains/premium-domain-purchases` feature flag to protect the functionality for the time being. The functional additions in these changes are also dependent on the related back end changes being developed in D46578-code, but should have no impact if the feature flag is disabled.

The actual changes that are implemented in this PR are:
 * Add the `domains/premium-domain-purchases` feature flag
 * Introduce the option of requesting premium suggestions, and limit that to cases where we are not in signup AND the feature flag is enabled
 * Add logic to display suggestion-specific pricing for premium suggestions
 * Update the availability checking logic to allow premium domains to be selected if the suggestions data flagged them as premium (i.e. we displayed appropriate price information)
 * -Update the checks for manually typed in domains to follow a similar pattern- - I have deferred this to a follow-up diff. We need to tackle this problem in a data-driven way, so I don't want to bloat this PR and/or introduce something that can cause complications for other premium TLDs.
 * Allow premium suggestions to be displayed for domain-only searches via `wordpress.com/domains`
 * This change pre-emptively pulls in the change from #44398 to use a different suggestions vendor for the domain-only flow instead of the signup-specific suggestions vendor.

**NOTE:** These changes should not affect the existing onboarding flows at `wordpress.com/start` or `wordpress.com/new`.

#### Testing instructions

The testing here is a bit complex, because we have four flows that we should test, namely "old" onboarding from `/start`, Gutenboarding from `/new`, the domain-only flow from `/domains`, and the "Add a domain" from the domain management pages in Calypso. Furthermore, we also have a config/feature flag that should have no impact on the two onboarding flows, and a minor impact on the two domain-specific flows, depending on what terms you search for.

##### Tests with config flag off

For each of the four flows, verify that you can do the following:
 * Search for a domain -- all four flows should call `https://public-api.wordpress.com/rest/v1.1/domains/suggestions` and should specify `premium=0`. The two onboarding flows should supply a `vendor` value of `variation4_front`, while the Calypso and domain-only flows should supply a vendor value of `variation2_front`.
 * Make sure you do a search for a premium .blog name, and verify that you don't get any premium results with different prices. (A good bet for premium names are a common Anglophone name followed by ` blog`, e.g. `matthew blog`, `claire blog`, `anita blog`.)

#### Testing with config flag on

Ensure that you start each flow by loading the URL with a suffix of `?flags=domains/premium-domain-purchases` to enable the flag. Also note that you will need your back end to be running/patched with D46578-code.
 * For `/new` and `/start`, the behaviour should be identical to the tests above.
 * For `/domains` and Calypso, the search should call the same API as the tests above, but should pass `vendor=variation7_front` and `premium=1`.
   - Make sure you search for some premium .blog domains (e.g. `dale blog`, `jane blog`, `anita blog`) and that you get results with premium prices. Ensure that you can move through all the way to checkout with a premium domain, and verify that the price is kept. (In particular, you should not see a prompt that WordPress.com doesn't support premium domains.)
  - Try searching for premium domains under some other TLD -- you shouldn't see any premium results.
  - Try typing in a premium domain name into the search field, e.g. `dale.blog` -- you should see an error about premium domains not being supported.
